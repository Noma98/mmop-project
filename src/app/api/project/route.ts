import { NextRequest, NextResponse } from 'next/server';
import {
  createProjects,
  deleteProject,
  updateProject,
} from '@/service/project';
import { withSessionUser } from '@/utils/session';

export async function POST(req: NextRequest) {
  return withSessionUser(async () => {
    const form = await req.formData();
    const data = JSON.parse(form.get('data')?.toString() as string);
    const files = form.getAll('file');

    return createProjects({
      data,
      files,
    })
      .then((res) => NextResponse.json(res))
      .catch((err) => new Response(JSON.stringify(err), { status: 500 }));
  });
}
export async function PUT(req: NextRequest) {
  return withSessionUser(async () => {
    const form = await req.formData();
    const data = JSON.parse(form.get('data')?.toString() as string);
    const files = form.getAll('file');

    return updateProject({
      data,
      files,
    })
      .then((res) => NextResponse.json(res))
      .catch((err) => new Response(JSON.stringify(err), { status: 500 }));
  });
}
export async function DELETE(req: NextRequest) {
  return withSessionUser(async () => {
    const { id } = await req.json();

    return deleteProject(id)
      .then((res) => NextResponse.json(res))
      .catch((err) => new Response(JSON.stringify(err), { status: 500 }));
  });
}
