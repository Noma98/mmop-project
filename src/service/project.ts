import { nanoid } from 'nanoid';
import { format } from 'date-fns';

import { SanityImageSource } from '@sanity/image-url/lib/types/types';
import { client, extractAssetIdFromUrl, urlFor } from '@/service/sanity';

export type Project = {
  id: string;
  title: string;
  authorId: string;
  startDate: string;
  endDate: string;
  skills: string[];
  description: string;
  type: 'business' | 'side';
  images: string[];
  webLink: string;
  storeLink: string;
  githubLink: string;
  achievements: Array<{ value: string }>;
};
export type RawProject = Omit<Project, 'id'> & {
  images: SanityImageSource[];
  _id: string;
};
export const getProjects = async (userId: string) => {
  const data = await client.fetch(
    `*[_type=="project"&&authorId=="${userId}"]|order("endDate" desc)`
  );
  return data.map((item: RawProject) => ({
    ...item,
    id: item._id,
    images: item.images.map((imgObj: SanityImageSource) => urlFor(imgObj)),
  }));
};
type ProjectFormData = {
  data: Omit<Project, 'images'> & { imageUrls: string[] };
  files: FormDataEntryValue[];
};
export const createProjects = async ({
  data: {
    achievements,
    authorId,
    description,
    endDate,
    githubLink,
    id,
    skills,
    startDate,
    storeLink,
    title,
    type,
    webLink,
  },
  files,
}: ProjectFormData) => {
  const imageAssets = await Promise.all(
    files.map((file) => client.assets.upload('image', file))
  );
  const imageAssetIds = imageAssets.map((imageAsset) => imageAsset._id);
  return client.create({
    _type: 'project',
    title,
    description,
    achievements: achievements.map((item) => ({ ...item, _key: nanoid() })),
    skills,
    githubLink,
    webLink,
    storeLink,
    images: imageAssetIds.map((id) => ({
      _key: nanoid(),
      asset: { _ref: id },
    })),
    startDate: format(new Date(startDate), 'yyyy-MM-dd'),
    endDate: format(new Date(endDate), 'yyyy-MM-dd'),
    type,
    authorId,
  });
};
export const updateProject = async ({
  data: {
    achievements,
    authorId,
    description,
    endDate,
    githubLink,
    id,
    skills,
    startDate,
    storeLink,
    title,
    type,
    webLink,
    imageUrls,
  },
  files,
}: ProjectFormData) => {
  const imageAssets = await Promise.all(
    files.map((file) => client.assets.upload('image', file))
  );
  const existingImgArr = imageUrls.map((url) => {
    const assetId = extractAssetIdFromUrl(url);
    if (!assetId) {
      throw new Error(`Could not extract asset ID from URL:${url}`);
    }
    return {
      _key: nanoid(),
      _type: 'image',
      asset: {
        _type: 'reference',
        _ref: assetId,
      },
    };
  });
  const newImgArr = imageAssets.map((imageAsset) => ({
    _key: nanoid(),
    _type: 'image',
    asset: {
      _type: 'reference',
      _ref: imageAsset._id,
    },
  }));
  return client
    .patch(id)
    .set({
      _type: 'project',
      title,
      description,
      achievements: achievements.map((item) => ({ ...item, _key: nanoid() })),
      skills,
      githubLink,
      webLink,
      storeLink,
      images: [...existingImgArr, ...newImgArr],
      startDate: format(new Date(startDate), 'yyyy-MM-dd'),
      endDate: format(new Date(endDate), 'yyyy-MM-dd'),
      type,
      authorId,
    })
    .commit();
};
export const deleteProject = (id: string) => {
  return client.delete(id);
};
