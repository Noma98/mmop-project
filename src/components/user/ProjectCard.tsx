import Label from '@/components/common/Label';
import {
  BusinessIcon,
  GameIcon,
  GithubIcon,
  StoreIcon,
  WebIcon,
} from '@/components/icons';
import IconLink from '@/components/user/IconLink';
import Period from '@/components/user/Period';
import { Project } from '@/service/project';

type Props = {
  data: Project;
};
export default function ProjectCard({ data }: Props) {
  const {
    title,
    description,
    images,
    webLink,
    githubLink,
    storeLink,
    achievements,
    skills,
    startDate,
    endDate,
    type,
  } = data;
  const linkData = [
    { text: 'Web Link', link: webLink, color: 'red-500', icon: <WebIcon /> },
    {
      text: 'Store Link',
      link: storeLink,
      color: '[#1C95F1]',
      icon: <StoreIcon />,
    },
    { text: 'Github Link', link: githubLink, icon: <GithubIcon /> },
  ];
  return (
    <article className='flex flex-col lg:flex-row p-10 round relative rounded-xl gap-10 m-8 shadow-md border-[1px] border-neutral-200 bg-white z-10'>
      <Label
        text={type}
        icon={type === 'business' ? <BusinessIcon /> : <GameIcon />}
        style='absolute top-[-20px] left-[20px]'
      />
      <div className={`flex flex-col gap-2 lg:hidden`}>
        <Period startDate={startDate} endDate={endDate} />
        <h2 className='text-2xl font-bold'>{title}</h2>
        <p>{description}</p>
      </div>
      <div className='flex flex-col gap-2 flex-1'>
        <div className={`hidden flex-col gap-2 lg:flex`}>
          <Period startDate={startDate} endDate={endDate} />
          <h2 className='text-2xl font-bold'>{title}</h2>
          <p>{description}</p>
        </div>
        <ul>
          {achievements.map((item, idx) => (
            <li
              className='list-square translate-x-4 text-neutral-700'
              key={idx}
            >
              {item.value}
            </li>
          ))}
        </ul>
        <div className='flex flex-wrap gap-2 items-center border-t-[1px] pt-4 mt-4'>
          {linkData.map(
            (data, idx) => data.link && <IconLink key={idx} {...data} />
          )}
        </div>
        <div className='flex gap-2 flex-wrap'>
          {skills.map((skill, idx) => (
            <span
              key={idx}
              className='py-1 px-2 rounded-lg text-sm bg-neutral-500 text-white font-bold'
            >
              {skill}
            </span>
          ))}
        </div>
      </div>
    </article>
  );
}
