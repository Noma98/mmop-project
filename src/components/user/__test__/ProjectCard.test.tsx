import { ReactNode } from 'react';
import { render, screen } from '@testing-library/react';

import ProjectCard from '@/components/user/ProjectCard';
import { testProject } from '@/tests/project';

type Props = {
  children: ReactNode;
};
jest.mock(
  '@/components/user/ImageCarousel',
  jest.fn(() => ({ children }: Props) => (
    <div data-testid='carousel'>{children}</div>
  ))
);

describe('ProjectCard', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<ProjectCard data={testProject} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders business lable if project type is business ', () => {
    render(<ProjectCard data={{ ...testProject, type: 'business' }} />);
    expect(screen.getByText('business')).toBeInTheDocument();
  });
  it('renders side label if project type is side', () => {
    render(<ProjectCard data={{ ...testProject, type: 'side' }} />);
    expect(screen.getByText('side')).toBeInTheDocument();
  });
  it('renders image carousel if images length is 1 or higher ', () => {
    render(<ProjectCard data={testProject} />);
    expect(screen.getByTestId('carousel')).toBeInTheDocument();
    expect(screen.getAllByAltText(/Test Project image/)).toHaveLength(
      testProject.images.length
    );
  });
  it('renders empty element if images length is 0', () => {
    render(<ProjectCard data={{ ...testProject, images: [] }} />);
    expect(screen.queryByTestId('carousel')).not.toBeInTheDocument();
  });
  it('must render at least one achievement', () => {
    render(<ProjectCard data={testProject} />);
    expect(screen.getAllByRole('listitem').length).toBeGreaterThan(1);
  });
});
