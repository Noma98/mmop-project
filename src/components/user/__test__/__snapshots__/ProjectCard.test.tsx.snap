// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard renders correctly 1`] = `
<DocumentFragment>
  <article
    class="flex flex-col lg:flex-row p-10 round relative rounded-xl gap-4 md:gap-10 border-neutral-200 bg-white z-10"
  >
    <div
      class="py-2 px-4 text-white font-bold rounded-full items-center flex gap-1 absolute top-[-20px] left-[20px] bg-black"
    >
      <svg
        class="w-5 h-5"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M146 80h220c9.9 0 18-8.1 18-18s-8.1-18-18-18H146c-9.9 0-18 8.1-18 18s8.1 18 18 18zM398 100H114c-9.9 0-18 8.1-18 18 0 9.3 6.9 17.1 16 18v300c0 13.2 10.8 24 24 24h88c4.4 0 8-3.6 8-8v-40c0-4.4 3.6-8 8-8h32c4.4 0 8 3.6 8 8v40c0 4.4 3.6 8 8 8h88c13.2 0 24-10.8 24-24V136c9.1-.9 16-8.7 16-18 0-9.9-8.1-18-18-18zM192 364c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm80 192c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm80 192c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16zm0-64c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-16c0-4.4 3.6-8 8-8h16c4.4 0 8 3.6 8 8v16z"
        />
      </svg>
      business
    </div>
    <div
      class="flex flex-col gap-2 lg:hidden"
    >
      <div
        class="flex gap-2 text-neutral-500 text-sm"
      >
        <div
          class="flex gap-2 items-center"
        >
          <svg
            class="w-4 h-4"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
            />
          </svg>
          <time>
            2023-01
          </time>
        </div>
        ~
        <div
          class="flex gap-2 items-center"
        >
          <svg
            class="w-4 h-4"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
            />
          </svg>
          <time>
            2023-12
          </time>
        </div>
      </div>
      <h2
        class="text-2xl font-bold"
      >
        Test Project
      </h2>
      <p>
        This is a test project description.
      </p>
    </div>
    <div
      data-testid="carousel"
    >
      <div
        class="h-[300px] bg-neutral-100 relative"
      >
        <img
          alt="Test Project image"
          class="object-contain"
          data-nimg="fill"
          decoding="async"
          fetchpriority="high"
          sizes="300"
          src="/_next/image?url=%2Ftest1.png&w=3840&q=75"
          srcset="/_next/image?url=%2Ftest1.png&w=16&q=75 16w, /_next/image?url=%2Ftest1.png&w=32&q=75 32w, /_next/image?url=%2Ftest1.png&w=48&q=75 48w, /_next/image?url=%2Ftest1.png&w=64&q=75 64w, /_next/image?url=%2Ftest1.png&w=96&q=75 96w, /_next/image?url=%2Ftest1.png&w=128&q=75 128w, /_next/image?url=%2Ftest1.png&w=256&q=75 256w, /_next/image?url=%2Ftest1.png&w=384&q=75 384w, /_next/image?url=%2Ftest1.png&w=640&q=75 640w, /_next/image?url=%2Ftest1.png&w=750&q=75 750w, /_next/image?url=%2Ftest1.png&w=828&q=75 828w, /_next/image?url=%2Ftest1.png&w=1080&q=75 1080w, /_next/image?url=%2Ftest1.png&w=1200&q=75 1200w, /_next/image?url=%2Ftest1.png&w=1920&q=75 1920w, /_next/image?url=%2Ftest1.png&w=2048&q=75 2048w, /_next/image?url=%2Ftest1.png&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <div
        class="h-[300px] bg-neutral-100 relative"
      >
        <img
          alt="Test Project image"
          class="object-contain"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          sizes="300"
          src="/_next/image?url=%2Ftest2.png&w=3840&q=75"
          srcset="/_next/image?url=%2Ftest2.png&w=16&q=75 16w, /_next/image?url=%2Ftest2.png&w=32&q=75 32w, /_next/image?url=%2Ftest2.png&w=48&q=75 48w, /_next/image?url=%2Ftest2.png&w=64&q=75 64w, /_next/image?url=%2Ftest2.png&w=96&q=75 96w, /_next/image?url=%2Ftest2.png&w=128&q=75 128w, /_next/image?url=%2Ftest2.png&w=256&q=75 256w, /_next/image?url=%2Ftest2.png&w=384&q=75 384w, /_next/image?url=%2Ftest2.png&w=640&q=75 640w, /_next/image?url=%2Ftest2.png&w=750&q=75 750w, /_next/image?url=%2Ftest2.png&w=828&q=75 828w, /_next/image?url=%2Ftest2.png&w=1080&q=75 1080w, /_next/image?url=%2Ftest2.png&w=1200&q=75 1200w, /_next/image?url=%2Ftest2.png&w=1920&q=75 1920w, /_next/image?url=%2Ftest2.png&w=2048&q=75 2048w, /_next/image?url=%2Ftest2.png&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
    </div>
    <div
      class="flex flex-col gap-2 flex-1"
    >
      <div
        class="hidden flex-col gap-2 lg:flex"
      >
        <div
          class="flex gap-2 text-neutral-500 text-sm"
        >
          <div
            class="flex gap-2 items-center"
          >
            <svg
              class="w-4 h-4"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
              />
            </svg>
            <time>
              2023-01
            </time>
          </div>
          ~
          <div
            class="flex gap-2 items-center"
          >
            <svg
              class="w-4 h-4"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
              />
            </svg>
            <time>
              2023-12
            </time>
          </div>
        </div>
        <h2
          class="text-2xl font-bold"
        >
          Test Project
        </h2>
        <p>
          This is a test project description.
        </p>
      </div>
      <ul>
        <li
          class="list-square translate-x-4 text-neutral-700"
        >
          Achievement 1
        </li>
        <li
          class="list-square translate-x-4 text-neutral-700"
        >
          Achievement 2
        </li>
      </ul>
      <div
        class="flex flex-wrap gap-2 items-center border-t-[1px] pt-4 mt-4"
      >
        <a
          class="whitespace-nowrap py-2 px-4 rounded-2xl text-white font-bold flex gap-2 items-center bg-red-500"
          href="https://test.com"
          target="_blank"
        >
          <svg
            class="w-5 h-5"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm-5 14H4v-4h11v4zm0-5H4V9h11v4zm5 5h-4V9h4v9z"
            />
          </svg>
          Web Link
        </a>
        <a
          class="whitespace-nowrap py-2 px-4 rounded-2xl text-white font-bold flex gap-2 items-center bg-sky-500"
          href="https://store.com/test"
          target="_blank"
        >
          <svg
            class="h-5 w-5"
            fill="currentColor"
            height="1em"
            role="img"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.8086 14.9194l6.1107-11.0368c.0837-.1513.1682-.302.2437-.4584.0685-.142.1267-.2854.1646-.4403.0803-.3259.0588-.6656-.066-.9767-.1238-.3095-.3417-.5678-.6201-.7355a1.4175 1.4175 0 0 0-.921-.1924c-.3207.043-.6135.1935-.8443.4288-.1094.1118-.1996.2361-.2832.369-.092.1463-.175.2979-.259.4492l-.3864.6979-.3865-.6979c-.0837-.1515-.1667-.303-.2587-.4492-.0837-.1329-.1739-.2572-.2835-.369-.2305-.2353-.5233-.3857-.844-.429a1.4181 1.4181 0 0 0-.921.1926c-.2784.1677-.4964.426-.6203.7355-.1246.311-.1461.6508-.066.9767.038.155.0962.2984.1648.4403.0753.1564.1598.307.2437.4584l1.248 2.2543-4.8625 8.7825H2.0295c-.1676 0-.3351-.0007-.5026.0092-.1522.009-.3004.0284-.448.0714-.3108.0906-.5822.2798-.7783.548-.195.2665-.3006.5929-.3006.9279 0 .3352.1057.6612.3006.9277.196.2683.4675.4575.7782.548.1477.043.296.0623.4481.0715.1675.01.335.009.5026.009h13.0974c.0171-.0357.059-.1294.1-.2697.415-1.4151-.6156-2.843-2.0347-2.843zM3.113 18.5418l-.7922 1.5008c-.0818.1553-.1644.31-.2384.4705-.067.1458-.124.293-.1611.452-.0785.3346-.0576.6834.0645 1.0029.1212.3175.3346.583.607.7549.2727.172.5891.2416.9013.1975.3139-.044.6005-.1986.8263-.4402.1072-.1148.1954-.2424.2772-.3787.0902-.1503.1714-.3059.2535-.4612L6 19.4636c-.0896-.149-.9473-1.4704-2.887-.9218m20.5861-3.0056a1.4707 1.4707 0 0 0-.779-.5407c-.1476-.0425-.2961-.0616-.4483-.0705-.1678-.0099-.3352-.0091-.503-.0091H18.648l-4.3891-7.817c-.6655.7005-.9632 1.485-1.0773 2.1976-.1655 1.0333.0367 2.0934.546 3.0004l5.2741 9.3933c.084.1494.167.299.2591.4435.0837.131.1739.2537.2836.364.231.2323.5238.3809.8449.4232.3192.0424.643-.0244.9217-.1899.2784-.1653.4968-.4204.621-.7257.1246-.3072.146-.6425.0658-.9641-.0381-.1529-.0962-.2945-.165-.4346-.0753-.1543-.1598-.303-.2438-.4524l-1.216-2.1662h1.596c.1677 0 .3351.0009.5029-.009.1522-.009.3007-.028.4483-.0705a1.4707 1.4707 0 0 0 .779-.5407A1.5386 1.5386 0 0 0 24 16.452a1.539 1.539 0 0 0-.3009-.9158Z"
            />
          </svg>
          Store Link
        </a>
        <a
          class="whitespace-nowrap py-2 px-4 rounded-2xl text-white font-bold flex gap-2 items-center bg-neutral-800"
          href="https://github.com/test"
          target="_blank"
        >
          <svg
            class="w-5 h-5"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 496 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
            />
          </svg>
          Github Link
        </a>
      </div>
      <div
        class="flex gap-2 flex-wrap"
      >
        <span
          class="py-1 px-2 rounded-lg text-sm bg-neutral-500 text-white font-bold"
        >
          React
        </span>
        <span
          class="py-1 px-2 rounded-lg text-sm bg-neutral-500 text-white font-bold"
        >
          JavaScript
        </span>
        <span
          class="py-1 px-2 rounded-lg text-sm bg-neutral-500 text-white font-bold"
        >
          CSS
        </span>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
